internal/controller/controllers/clusterdeployments_controller_test.go

```diff
@@ -4198,6 +4207,54 @@ var _ = Describe("unbindAgents", func() {
 		err = c.Get(ctx, types.NamespacedName{Name: "test-agent2", Namespace: testNamespace}, agent)
 		Expect(k8serrors.IsNotFound(err)).To(BeTrue())
 	})
+
+	It("deletes late bound agents when cluster deployment is preserved", func() {
+		infraEnv := &aiv1beta1.InfraEnv{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "InfraEnv",
+				APIVersion: fmt.Sprintf("%s/%s", aiv1beta1.GroupVersion.Group, aiv1beta1.GroupVersion.Version),
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      infraEnvName,
+				Namespace: testNamespace,
+			},
+			Spec: aiv1beta1.InfraEnvSpec{ClusterRef: nil},
+		}
+		Expect(c.Create(ctx, infraEnv)).To(Succeed())
+
+		clusterDeployment := &hivev1.ClusterDeployment{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: clusterReference.Namespace,
+				Name:      clusterReference.Name,
+			},
+			Spec: hivev1.ClusterDeploymentSpec{
+				PreserveOnDelete: true,
+			},
+		}
+		Expect(c.Create(ctx, clusterDeployment)).To(Succeed())
```

Comment:
Instead of recreating these could you use the existing ones already defined in the test setup?

Reply:
I don't see this created by the test setup, rather individual tests create it explicitly. I can change this to use the `newClusterDeployment` function, if that is what you mean.

Reply:
It would be something like this, which isn't much better:

```diff
diff --git a/internal/controller/controllers/clusterdeployments_controller_test.go b/internal/controller/controllers/clusterdeployments_controller_test.go
index ed612a28e..ba71eaba9 100644
--- a/internal/controller/controllers/clusterdeployments_controller_test.go
+++ b/internal/controller/controllers/clusterdeployments_controller_test.go
@@ -4218,15 +4218,13 @@ var _ = Describe("unbindAgents", func() {
                }
                Expect(c.Create(ctx, infraEnv)).To(Succeed())
 
-               clusterDeployment := &hivev1.ClusterDeployment{
-                       ObjectMeta: metav1.ObjectMeta{
-                               Namespace: clusterReference.Namespace,
-                               Name:      clusterReference.Name,
-                       },
-                       Spec: hivev1.ClusterDeploymentSpec{
+               clusterDeployment := newClusterDeployment(
+                       clusterReference.Name,
+                       clusterReference.Namespace,
+                       hivev1.ClusterDeploymentSpec{
                                PreserveOnDelete: true,
                        },
-               }
+               )
                Expect(c.Create(ctx, clusterDeployment)).To(Succeed())
 
                agent := &aiv1beta1.Agent{
```

Reply:
Ah sorry I misread the setup. I guess it's only creating the agent. What you have is fine then. You could use one of the agents that was created in the setup, but it's not a big deal.

Reply:
Done, the test uses now the agents created in the setup.